### DISCUSSION TITLE
How to create fixture for sealed class with no public constructor.

### DISCUSSION THREAD

**Discussion by JayramJai** [AutoFixture/AutoFixture#1264] [Q&A]
I have sealed the class with no constructor which is I am referred using SDK to my project. I want to create fixture data for the class to write test but AutoFixture is giving an expectation like below.

'AutoFixture was unable to create an instance from `SealedTypeclass`, most likely because it has no public constructor, is an abstract or non-public type.

Any generic solution?

**Comment by aivascu** [comment]
There is very little information about the problem you're trying to solve.
Could you provide an example of how you would write the test without AutoFixture?

**Comment by JayramJai** [comment]
@aivascu Please find the below code example for which I am trying to create a fixture.


    public sealed class TokenCacheItem
    {
        public string Authority { get; }
        public string ClientId { get; }
        public DateTimeOffset ExpiresOn { get; }
        public string FamilyName { get; }
        public string GivenName { get; }
        public string IdentityProvider { get; }
        public string TenantId { get; }
}

With the above-sealed class i am referring through SDK and I am trying to create fixture data. I am getting the below error message. 

**Message:** 
    AutoFixture.ObjectCreationExceptionWithPath : AutoFixture was unable to create an instance from ***, most likely because it has no public constructor, is an abstract or non-public type.

Is it sufficient?

**Comment by aivascu** [comment]
@JayramJai as you can see in this [example](https://dotnetfiddle.net/b5kpf1), AutoFixture has no problems creating an instance of the class you provided.

I am assuming you also made the constructor private or protected. AutoFixture cannot create an instance of a class that has no public constructor (also if the only public constructor is a copy constructor) and neither it can fill properties with a non public setter.
Basically if you cannot construct manually a type instance you cannot expect AutoFixture to be able to.

You mentioned you create instances using an SDK. Perhaps you could customize your types to be constructed using a factory that uses the said SDK.

```cs
fixture.Customize<TokenCacheItem>(
    c => c.FromFactory(() => SdkFactory.Create()));
```

**Reply by JayramJai** [reply]
@aivascu What you mean by SdkFactory?

**Reply by aivascu** [reply]
@JayramJai if I understood you correctly, you were saying that `TokenCacheItem` instances are created inside an SDK, i.e. a library that can access the non-public constructor of `TokenCacheItem`. `SdkFactory` is supposed to represent an abstraction from your SDK, that creates  `TokenCacheItem` instances, assuming something similar exists in your SDK.
